app/src/main/java/org/wikipedia/diff/ArticleEditDetailsFragment.kt

```diff
@@ -0,0 +1,401 @@
+                            R.attr.material_theme_de_emphasised_color))
+                    thankButton.isClickable = false
+                })
+                { Consumer { t: Throwable? -> L.e(t) } }
+    }
+
+    private fun fetchDiffText() {
+        disposables.add(ServiceFactory.getCoreRest(WikiSite.forLanguageCode(languageCode)).getDiff(olderRevisionId, revisionId)
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe({
+                    displayDiff(it.diffs)
+                }) { t: Throwable? -> L.e(t) })
+    }
+
+    private fun displayDiff(diffs: MutableList<DiffItem>) {
+        for (diff in diffs) {
+            val prefixLength = diffText.text.length
+            val spannableString = SpannableStringBuilder(diffText.text).append(if (diff.text.isNotEmpty()) diff.text else "\n")
+            when (diff.type) {
+                DIFF_TYPE_LINE_WITH_SAME_CONTENT -> {
+                }
+                DIFF_TYPE_LINE_ADDED -> {
+                    updateDiffTextDecor(spannableString, true, prefixLength, prefixLength + diff.text.length)
+                }
+                DIFF_TYPE_LINE_REMOVED -> {
+                    updateDiffTextDecor(spannableString, false, prefixLength, prefixLength + diff.text.length)
+                }
+                DIFF_TYPE_LINE_WITH_DIFF -> {
+                    for (hightlightRange in diff.highlightRanges) {
+                        val highlightRangeStart = if (languageCode == "en") hightlightRange.start
+                        else getByteInCharacters(diff.text, hightlightRange.start, 0)
+                        val highlightRangeLength = if (languageCode == "en") hightlightRange.length
+                        else getByteInCharacters(diff.text, hightlightRange.length, highlightRangeStart)
+
+                        if (hightlightRange.type == HIGHLIGHT_TYPE_ADD) {
+                            updateDiffTextDecor(spannableString, true, prefixLength + highlightRangeStart, prefixLength + highlightRangeStart + highlightRangeLength)
+                        } else {
+                            updateDiffTextDecor(spannableString, false, prefixLength + highlightRangeStart, prefixLength + highlightRangeStart + highlightRangeLength)
+                        }
+                    }
+                }
+                DIFF_TYPE_PARAGRAPH_MOVED_FROM -> {
+                    updateDiffTextDecor(spannableString, false, prefixLength, prefixLength + diff.text.length)
+                }
+                DIFF_TYPE_PARAGRAPH_MOVED_TO -> {
+                    updateDiffTextDecor(spannableString, true, prefixLength, prefixLength + diff.text.length)
+                }
+            }
+            diffText.text = spannableString.append("\n")
+        }
+    }
+
+    private fun updateDiffTextDecor(spannableText: SpannableStringBuilder, isAddition: Boolean, start: Int, end: Int) {
```

Comment:
This is much better now 👍 

I noticed that the texts are not aligning to the center vertically (in the diff background color), can we optimize it?

Reply:
OK i finally realized this is because of the lineSpacingExtra attribute... I will try to see if there is a work around.

Reply:
Hi @sharvaniharan, please have a follow-up PR if you can resolve the issue. I'll merge this PR to make things move.